src/common/io/tls/session.c

```diff
@@ -75,52 +75,92 @@ tlsSessionClose(TlsSession *this, bool shutdown)
 ***********************************************************************************************************************************/
-static bool
-tlsSessionError(TlsSession *this, int code)
+// Helper to process error conditions
+static int
+tlsSessionResultProcess(TlsSession *this, int errorTls, int errorSys, bool closeOk)
 {
     FUNCTION_LOG_BEGIN(logLevelTrace);
         FUNCTION_LOG_PARAM(TLS_SESSION, this);
-        FUNCTION_LOG_PARAM(INT, code);
+        FUNCTION_LOG_PARAM(INT, errorTls);
+        FUNCTION_LOG_PARAM(INT, errorSys);
+        FUNCTION_LOG_PARAM(BOOL, closeOk);
     FUNCTION_LOG_END();
 
-    bool result = false;
+    ASSERT(this != NULL);
+    ASSERT(this->session != NULL);
+
+    int result = -1;
 
-    switch (code)
+    switch (errorTls)
     {
         // The connection was closed
         case SSL_ERROR_ZERO_RETURN:
         {
+            if (!closeOk)
+                THROW(ProtocolError, "unexpected TLS eof");
+
             tlsSessionClose(this, false);
             break;
         }
 
-        // Try the read/write again
+        // Try again after waiting for read ready
         case SSL_ERROR_WANT_READ:
-        case SSL_ERROR_WANT_WRITE:
         {
-            result = true;
+            sckSessionReadyRead(this->socketSession);
+            result = 0;
             break;
         }
 
-        // A syscall failed (usually indicates unexpected eof)
-        case SSL_ERROR_SYSCALL:
+        // Try again after waiting for write ready
+        case SSL_ERROR_WANT_WRITE:
         {
-            // Get the error before closing so it is not cleared
-            int errNo = errno;
-            tlsSessionClose(this, false);
-
-            // Throw the sys error
-            THROW_SYS_ERROR_CODE(errNo, KernelError, "tls failed syscall");
+            sckSessionReadyWrite(this->socketSession);
+            result = 0;
+            break;
         }
 
-        // Some other tls error that cannot be handled
+        // A syscall failed (this usually indicates unexpected eof)
+        case SSL_ERROR_SYSCALL:
+            THROW_SYS_ERROR_CODE(errorSys, KernelError, "TLS syscall error");
+
+        // Any other error that we cannot handle
         default:
-            THROW_FMT(ServiceError, "tls error [%d]", code);
+            THROW_FMT(ServiceError, "TLS error [%d]", errorTls);
+    }
+
+    FUNCTION_LOG_RETURN(INT, result);
+}
+
+static int
+tlsSessionResult(TlsSession *this, int result, bool closeOk)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(TLS_SESSION, this);
+        FUNCTION_LOG_PARAM(INT, result);
+        FUNCTION_LOG_PARAM(BOOL, closeOk);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(this->session != NULL);
+
+    // Process errors
+    if (result <= 0)
+    {
+        // Get TLS error and store errno in case of syscall error
+        int errorTls = SSL_get_error(this->session, result);
+        int errorSys = errno;
```

Comment:
Yeah -- it's fine for the SSL_error but probably not for errno. I tried capturing errno right after the SSL_read() but it was zero in all error cases that I could cause with testing. People have complained about that on the interwebs, too.

I think we can improve this but it doesn't change the program flow. There aren't any recoverable syscall errors.

Reply:
Also, it's not clear to. me when errno should be captured. PG does it after the SSL_get_error() and ERR_get_error() calls like here, but there's less in between.

Having said that, I don't think anything in our debug macros is resetting errno. We only do it in cvtZToUInt64Internal() and cvtZToInt64Internal(), which is not done in the debug macros.

Reply:
I was specifically worried that some of those macros might expand in certain cases to do things like snprintf like in typeToLog() which can set errno on some platforms.

The PG code does SSL_write, SSL_get_error, then calls ERR_get_error if SSL_write returned < 0 OR if SSL_get_error returned anything other than SSL_ERROR_NONE, and then looks at errno only if SSL_ERROR_SYSCALL was returned from SSL_get_error.

It seems like the SSL code is perhaps managing its own queue of errno results and setting it only when SSL_ERROR_SYSCALL is returned from SSL_get_error (and that's probably the only time we should actually be looking at errno after SSL_get_error...  maybe we should also explicit set errno to 0 before calling any of the above, like the PG code does), but I have to admit that I'm still not thrilled at having so much happening between these calls, which is quite unlike the PG code..

Reply:
I haven't been able to get any joy from ERR_get_error() either, but it may have just been the test cases I was trying. I tried a bunch of different ways to get some value from errno while trying to debug this issue.

Agreed, this is area for improvement, but I don't think we should do it now. Historically the value of this information has been pretty much NIL and it's only available at debug level so it is mostly never seen. Errors from S3 are pretty common so we don't log them at a higher level.

Reply:
Also, I do hope `snprintf` is not resetting errno on success. If that's happening on some platforms then they won't get detailed info in the error, not that we do on any platform as far as I can tell.